File: vortexdbg-android/src/main/java/com/vortexdbg/app/VortexSession.java
package com.vortexdbg.app;

import com.vortexdbg.AndroidEmulator;
import com.vortexdbg.arm.backend.BackendFactory;
import com.vortexdbg.linux.android.AndroidEmulatorBuilder;
import com.vortexdbg.linux.android.AndroidResolver;
import com.vortexdbg.linux.android.dvm.DalvikModule;
import com.vortexdbg.linux.android.dvm.DvmClass;
import com.vortexdbg.linux.android.dvm.VM;
import com.vortexdbg.linux.android.dvm.jni.ProxyClassFactory;
import com.vortexdbg.memory.Memory;

import java.io.Closeable;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Fachada de orquestração do Vortex-DBG (A1) — consolida os spikes WF1..Capstone numa
 * API única e usável.
 *
 * Um {@code VortexSession} monta, num único processo hermético e off-device:
 *   - o emulador nativo (UniDBG, backend Unicorn2 por padrão);
 *   - a Dalvik VM com a ponte JNI (propagação de exceção ligada);
 *   - o {@link VortexClassLoader} com as classes REAIS do app (extraídas via JEB/dex2jar)
 *     + a camada de framework ({@link VortexFramework} / android-all);
 *   - o {@link ProxyClassFactory} apontado a esse classloader, para que o FindClass/
 *     CallMethod do código nativo resolva as classes reais do app;
 *   - as bibliotecas nativas (.so) do app.
 *
 * Uso típico:
 * <pre>
 *   try (VortexSession s = VortexSession.builder()
 *           .classes(new File("app-classes.jar"))
 *           .androidAll(new File("android-all.jar"))
 *           .nativeLib(new File("libfoo.so"))
 *           .open()) {
 *       Object r = s.invokeStatic("com.app.Crypto", "decrypt",
 *               new Class[]{String.class}, "deadbeef");
 *   }
 * </pre>
 */
public class VortexSession implements Closeable {

    private final AndroidEmulator emulator;
    private final VM vm;
    private final VortexClassLoader classLoader;
    private final VortexInvoker invoker;
    private final List<DalvikModule> nativeModules;

    private VortexSession(AndroidEmulator emulator, VM vm, VortexClassLoader classLoader,
                          List<DalvikModule> nativeModules) {
        this.emulator = emulator;
        this.vm = vm;
        this.classLoader = classLoader;
        this.invoker = new VortexInvoker(classLoader);
        this.nativeModules = nativeModules;
    }

    public static Builder builder() {
        return new Builder();
    }

    // ---- API pós-open ----

    public AndroidEmulator emulator() { return emulator; }
    public VM vm() { return vm; }
    public VortexClassLoader classLoader() { return classLoader; }
    public VortexInvoker invoker() { return invoker; }
    public List<DalvikModule> nativeModules() { return nativeModules; }

    /** Carrega uma classe do app na JVM host. */
    public Class<?> loadAppClass(String binaryName) {
        return classLoader.loadApp(binaryName);
    }

    /** Invoca um método estático de uma classe do app (estilo LSPosed, off-device). */
    public Object invokeStatic(String className, String method, Class<?>[] paramTypes, Object... args) throws Exception {
        return invoker.invokeStatic(className, method, paramTypes, args);
    }

    /** Resolve uma DvmClass (nome JNI com '/') para chamadas vindas do lado nativo. */
    public DvmClass resolveNativeClass(String jniName) {
        return vm.resolveClass(jniName);
    }

    @Override
    public void close() {
        try {
            emulator.close();
        } catch (Exception ignored) {
            // best-effort
        }
    }

    // ---- Builder ----

    public static class Builder {
        private final List<File> classes = new ArrayList<>();
        private final List<File> nativeLibs = new ArrayList<>();
        private File androidAll;
        private int sdk = 23;
        private boolean is64 = true;
        private boolean verbose = false;
        private boolean exceptionPropagation = true;
        private boolean callJniOnLoad = false;
        private BackendFactory backendFactory;
        private String processName = "vortex";
        private File rootDir = new File("target/rootfs");

        public Builder classes(File... jarsOrDirs) {
            if (jarsOrDirs != null) {
                for (File f : jarsOrDirs) if (f != null) classes.add(f);
            }
            return this;
        }

        public Builder androidAll(File jar) { this.androidAll = jar; return this; }
        public Builder nativeLib(File... so) {
            if (so != null) for (File f : so) if (f != null) nativeLibs.add(f);
            return this;
        }
        public Builder sdk(int sdk) { this.sdk = sdk; return this; }
        public Builder arch64(boolean is64) { this.is64 = is64; return this; }
        public Builder verbose(boolean verbose) { this.verbose = verbose; return this; }
        public Builder exceptionPropagation(boolean enabled) { this.exceptionPropagation = enabled; return this; }
        public Builder callJniOnLoad(boolean enabled) { this.callJniOnLoad = enabled; return this; }
        public Builder backend(BackendFactory factory) { this.backendFactory = factory; return this; }
        public Builder processName(String name) { this.processName = name; return this; }
        public Builder rootDir(File dir) { this.rootDir = dir; return this; }

        public VortexSession open() {
            AndroidEmulatorBuilder eb = is64 ? AndroidEmulatorBuilder.for64Bit() : AndroidEmulatorBuilder.for32Bit();
            eb.setProcessName(processName);
            eb.setRootDir(rootDir);
            eb.addBackendFactory(backendFactory != null ? backendFactory : defaultBackend());
            AndroidEmulator emulator = eb.build();

            Memory memory = emulator.getMemory();
            memory.setLibraryResolver(new AndroidResolver(sdk));

            VM vm = emulator.createDalvikVM();
            vm.setVerbose(verbose);
            vm.setExceptionPropagation(exceptionPropagation);

            VortexClassLoader cl;
            if (androidAll != null) {
                cl = VortexFramework.fromAndroidAll(androidAll).newAppClassLoader(classes.toArray(new File[0]));
            } else {
                cl = new VortexClassLoader(classes.toArray(new File[0]));
            }
            vm.setDvmClassFactory(new ProxyClassFactory(cl));

            List<DalvikModule> modules = new ArrayList<>();
            for (File so : nativeLibs) {
                DalvikModule dm = vm.loadLibrary(so, callJniOnLoad);
                if (callJniOnLoad) {
                    dm.callJNI_OnLoad(emulator);
                }
                modules.add(dm);
            }

            return new VortexSession(emulator, vm, cl, modules);
        }

        private static BackendFactory defaultBackend() {
            try {
                Class<?> c = Class.forName("com.vortexdbg.arm.backend.Unicorn2Factory");
                return (BackendFactory) c.getConstructor(boolean.class).newInstance(true);
            } catch (Exception e) {
                throw new IllegalStateException(
                        "backend padrão (Unicorn2Factory) indisponível no classpath; use .backend(...)", e);
            }
        }
    }
}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/app/VortexSession.kt
package com.vortexdbg.app

import com.vortexdbg.AndroidEmulator
import com.vortexdbg.arm.backend.BackendFactory
import com.vortexdbg.linux.android.AndroidEmulatorBuilder
import com.vortexdbg.linux.android.AndroidResolver
import com.vortexdbg.linux.android.dvm.DalvikModule
import com.vortexdbg.linux.android.dvm.DvmClass
import com.vortexdbg.linux.android.dvm.VM
import com.vortexdbg.linux.android.dvm.jni.ProxyClassFactory
import java.io.Closeable
import java.io.File

/**
 * Orchestration facade for Vortex-DBG (A1) — wires the moving parts into a single, usable API.
 *
 * A [VortexSession] assembles, in one hermetic off-device process:
 *   - the native emulator (UniDBG, Unicorn2 backend by default);
 *   - the Dalvik VM with the JNI bridge (exception propagation enabled);
 *   - the [VortexClassLoader] holding the app's REAL classes (extracted via JEB/dex2jar)
 *     plus the framework layer ([VortexFramework] / android-all);
 *   - a [ProxyClassFactory] pointed at that classloader, so FindClass/CallMethod from native
 *     code resolves against the app's real classes;
 *   - the app's native libraries (.so).
 *
 * Typical use:
 * ```
 *   try (VortexSession s = VortexSession.builder()
 *           .classes(new File("app-classes.jar"))
 *           .androidAll(new File("android-all.jar"))
 *           .nativeLib(new File("libfoo.so"))
 *           .open()) {
 *       Object r = s.invokeStatic("com.app.Crypto", "decrypt",
 *               new Class[]{String.class}, "deadbeef");
 *   }
 * ```
 */
open class VortexSession private constructor(
    private val emulator: AndroidEmulator,
    private val vm: VM,
    private val classLoader: VortexClassLoader,
    private val nativeModules: List<DalvikModule>
) : Closeable {

    private val invoker: VortexInvoker = VortexInvoker(classLoader)

    // ---- post-open API ----

    open fun emulator(): AndroidEmulator { return emulator }
    open fun vm(): VM { return vm }
    open fun classLoader(): VortexClassLoader { return classLoader }
    open fun invoker(): VortexInvoker { return invoker }
    open fun nativeModules(): List<DalvikModule> { return nativeModules }

    /** Loads an app class into the host JVM. */
    open fun loadAppClass(binaryName: String): Class<*> {
        return classLoader.loadApp(binaryName)
    }

    /** Invokes a static method on an app class (LSPosed-style, off-device). */
    @Throws(Exception::class)
    open fun invokeStatic(className: String, method: String, paramTypes: Array<Class<*>>, vararg args: Any?): Any? {
        return invoker.invokeStatic(className, method, paramTypes, *args)
    }

    /** Resolves a DvmClass (JNI name with '/') for calls coming from the native side. */
    open fun resolveNativeClass(jniName: String): DvmClass {
        return vm.resolveClass(jniName)
    }

    override fun close() {
        try {
            emulator.close()
        } catch (ignored: Exception) {
            // best-effort: closing must not throw
        }
    }

    companion object {
        @JvmStatic
        fun builder(): Builder {
            return Builder()
        }
    }

    // ---- Builder ----

    open class Builder {
        private val classes = ArrayList<File>()
        private val nativeLibs = ArrayList<File>()
        private var androidAll: File? = null
        private var sdk = 23
        private var is64 = true
        private var verbose = false
        private var exceptionPropagation = true
        private var callJniOnLoad = false
        private var backendFactory: BackendFactory? = null
        private var processName = "vortex"
        private var rootDir: File = File("target/rootfs")

        open fun classes(vararg jarsOrDirs: File?): Builder {
            for (f in jarsOrDirs) if (f != null) classes.add(f)
            return this
        }

        open fun androidAll(jar: File?): Builder { this.androidAll = jar; return this }
        open fun nativeLib(vararg so: File?): Builder {
            for (f in so) if (f != null) nativeLibs.add(f)
            return this
        }
        open fun sdk(sdk: Int): Builder { this.sdk = sdk; return this }
        open fun arch64(is64: Boolean): Builder { this.is64 = is64; return this }
        open fun verbose(verbose: Boolean): Builder { this.verbose = verbose; return this }
        open fun exceptionPropagation(enabled: Boolean): Builder { this.exceptionPropagation = enabled; return this }
        open fun callJniOnLoad(enabled: Boolean): Builder { this.callJniOnLoad = enabled; return this }
        open fun backend(factory: BackendFactory?): Builder { this.backendFactory = factory; return this }
        open fun processName(name: String): Builder { this.processName = name; return this }
        open fun rootDir(dir: File): Builder { this.rootDir = dir; return this }

        open fun open(): VortexSession {
            val eb: AndroidEmulatorBuilder = if (is64) AndroidEmulatorBuilder.for64Bit() else AndroidEmulatorBuilder.for32Bit()
            eb.setProcessName(processName)
            eb.setRootDir(rootDir)
            eb.addBackendFactory(backendFactory ?: defaultBackend())
            val emulator = eb.build()

            val memory = emulator.getMemory()
            memory.setLibraryResolver(AndroidResolver(sdk))

            val vm = emulator.createDalvikVM()
            vm.setVerbose(verbose)
            vm.setExceptionPropagation(exceptionPropagation)

            val cl: VortexClassLoader
            if (androidAll != null) {
                cl = VortexFramework.fromAndroidAll(androidAll).newAppClassLoader(*classes.toTypedArray())
            } else {
                cl = VortexClassLoader(*classes.toTypedArray())
            }
            vm.setDvmClassFactory(ProxyClassFactory(cl))

            val modules = ArrayList<DalvikModule>()
            for (so in nativeLibs) {
                val dm = vm.loadLibrary(so, callJniOnLoad)
                if (callJniOnLoad) {
                    dm.callJNI_OnLoad(emulator)
                }
                modules.add(dm)
            }

            return VortexSession(emulator, vm, cl, modules)
        }

        companion object {
            private fun defaultBackend(): BackendFactory {
                try {
                    val c = Class.forName("com.vortexdbg.arm.backend.Unicorn2Factory")
                    return c.getConstructor(Boolean::class.javaPrimitiveType).newInstance(true) as BackendFactory
                } catch (e: Exception) {
                    throw IllegalStateException(
                        "backend padrão (Unicorn2Factory) indisponível no classpath; use .backend(...)", e
                    )
                }
            }
        }
    }
}
